src/VisualStudio/Core/Def/DesignerAttribute/VisualStudioDesignerAttributeService.cs

```diff
@@ -111,79 +133,180 @@ private async ValueTask ProcessWorkspaceChangeAsync(CancellationToken cancellati
+            using var _ = PooledHashSet<DocumentId>.GetInstance(out var seenDocuments);
 
-            await client.TryInvokeAsync<IRemoteDesignerAttributeDiscoveryService>(
-                solution,
-                (service, checksum, callbackId, cancellationToken) => service.DiscoverDesignerAttributesAsync(callbackId, checksum, priorityDocument, cancellationToken),
-                callbackTarget: this,
-                cancellationToken).ConfigureAwait(false);
+            // get the results and add all the documents we hear about to the notification queue so they can be batched up.
+            await foreach (var data in stream.ConfigureAwait(false))
+            {
+                seenDocuments.Add(data.DocumentId);
+                _projectSystemNotificationQueue.AddWork(data);
+            }
+
+            // now, if we didn't hear about some document that we previous reported, then go through and notify the project system
+            // that the document is no longer designable.
+            foreach (var (documentId, lastReportedData) in _documentToLastReportedData)
+            {
+                if (documentId.ProjectId == projectId &&
+                    !seenDocuments.Contains(documentId))
+                {
+                    _projectSystemNotificationQueue.AddWork(lastReportedData.WithCategory(null));
+                }
+            }
         }
 
         private async ValueTask NotifyProjectSystemAsync(
-            ImmutableSegmentedList<DesignerAttributeData> data, CancellationToken cancellationToken)
+            ImmutableSegmentedList<DesignerAttributeData> dataList, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            using var _1 = ArrayBuilder<DesignerAttributeData>.GetInstance(out var filteredInfos);
-            AddFilteredInfos(data, filteredInfos);
+            using var _1 = ArrayBuilder<DesignerAttributeData>.GetInstance(out var filteredData);
+            AddFilteredData(dataList, filteredData);
 
             // Now, group all the notifications by project and update all the projects in parallel.
             using var _2 = ArrayBuilder<Task>.GetInstance(out var tasks);
-            foreach (var group in filteredInfos.GroupBy(a => a.DocumentId.ProjectId))
+            foreach (var group in filteredData.GroupBy(a => a.DocumentId.ProjectId))
             {
                 cancellationToken.ThrowIfCancellationRequested();
                 tasks.Add(NotifyProjectSystemAsync(group.Key, group, cancellationToken));
             }
 
             // Wait until all project updates have happened before processing the next batch.
             await Task.WhenAll(tasks).ConfigureAwait(false);
+
+            // now that we've reported this data, record that we've done so so that we don't report the same data again in the future.
+            foreach (var data in filteredData)
+                _documentToLastReportedData[data.DocumentId] = data;
```

Comment:
I think this can race with the code in `ProcessResultsDataAsync` - take the scenario
1.  `ProcessResultsDataAsync` sees DocumentA has a designer attribute.
2.  The designer attribute is removed causing a second `ProcessResultsDataAsync` **before `NotifyProjectSystemAsync` runs for 1)**.  Since we haven't processed the notification to the project system for 1) yet, it doesn't know that we've reported a designer attribute for DocumentA and therefore does not add a null category result to the project system queue
3. The second queue eventually processes the first notification for 1) and DocumentA telling the project system there is a designer attribute in A
4.  The second queue doesn't have a notification for the removal in 2) because we hadn't done 3.  So we never report the removal.

Reply:
good call.  refactored.